File: merge_sort/python/merge_sort.py
from typing import List


def merge_sort(unsorted: List[int]) -> List[int]:
    if len(unsorted) <= 1:
        return unsorted

    mid = len(unsorted) // 2
    left = merge_sort(unsorted[:mid])
    right = merge_sort(unsorted[mid:])

    return merge(left, right)


def merge(left: List[int], right: List[int]) -> List[int]:
    result = []

    i = j = 0
    while i < len(left) and j < len(right):
        if left[i] < right[j]:
            result.append(left[i])
            i += 1
        else:
            result.append(right[j])
            j += 1

    result.extend(left[i:])
    result.extend(right[j:])
    return result

File: merge_sort/java/src/main/java/com/okayhak/MergeSort.java
package com.okayhak;

import java.util.ArrayList;
import java.util.List;

public class MergeSort {
    public static List<Integer> sort(List<Integer> unsortedArray) {
        if (unsortedArray.size() <= 1) {
            return unsortedArray;
        }

        int mid = unsortedArray.size() / 2;
        List<Integer> left = sort(unsortedArray.subList(0, mid));
        List<Integer> right = sort(unsortedArray.subList(mid, unsortedArray.size()));

        return merge(left, right);
    }

    public static List<Integer> merge(List<Integer> left, List<Integer> right) {
        List<Integer> result = new ArrayList<>();
        int i = 0;
        int j = 0;

        while (i < left.size() && j < right.size()) {
            if (left.get(i) < right.get(j)) {
                result.add(left.get(i));
                i++;
            } else {
                result.add(right.get(j));
                j++;
            }
        }

        while (i < left.size()) {
            result.add(left.get(i));
            i++;
        }

        while (j < right.size()) {
            result.add(right.get(j));
            j++;
        }

        return result;
    }
}
